### DISCUSSION TITLE
Is there a way to infer the resulting type of an Ast without evaluating it

### DISCUSSION THREAD

**Discussion by Natr1x** [PowerShell/PowerShell#21438] [Q&A] [answered]
I would like to write an argument completer for a cmdlet which completes on the property names of an InputObject (like `Select-Object -Property` does).

I figured I could use the CommandAst which is passed to the completer to find out what expressions precede the function in the pipeline but I can't seem to figure out how to then get the type that would result from this expression without evaluating it with `SafeEval`.

[System.Management.Automation/engine/parser/TypeInferenceVisitor.cs](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs) seems to be doing something like this and [System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs) has a `SafeExprEvaluator` class. But since both of those are internal I can't access them.

Or is there maybe some other way to find out type information of pipeline input in argument completers?

**Comment by MartinGC94** [comment, marked as answer]
I don't think you can do this without reflection, so you may as well use the Type Inference API rather than trying to reimplement it. Here's an example of how you can do it:
```
using namespace System.Management.Automation
using namespace System.Management.Automation.Language
$Method = [TypeInferenceRuntimePermissions].Assembly.GetType('System.Management.Automation.AstTypeInference').GetMethods() | Where-Object -FilterScript {
    $Params = $_.GetParameters()
    $Params.Count -eq 2 -and
    $Params[0].ParameterType -eq [Ast] -and
    $Params[1].ParameterType -eq [TypeInferenceRuntimePermissions]
}
$DemoAst = {Get-ChildItem C:\}.Ast
$InferredTypes = $Method.Invoke($null, ($DemoAst, [TypeInferenceRuntimePermissions]::AllowSafeEval))
```
Alternatively if you want to stick with public APIs you could parse the whole script to get the script text, replace the command at your cursor with `Select-Object` and then call `TabExpansion2` to get the completion results that PowerShell would have provided for `Select-Object`.
